File: backend/src/laboratorios/validators/date.utils.ts
export function parseDate(value: unknown): Date | null {
  if (!value) return null;
  if (value instanceof Date) return isNaN(value.getTime()) ? null : value;
  if (typeof value === 'number') {
    // Excel serial date: days since 1900-01-01 (with Lotus 1-2-3 bug for day 60)
    const utc_days = Math.floor(value - 25569);
    const date = new Date(utc_days * 86400000);
    return isNaN(date.getTime()) ? null : date;
  }
  if (typeof value === 'string') {
    const trimmed = value.trim();
    // dd/mm/aaaa
    const ddmmyyyy = /^(\d{1,2})\/(\d{1,2})\/(\d{4})$/.exec(trimmed);
    if (ddmmyyyy) {
      const [, d, m, y] = ddmmyyyy;
      const date = new Date(Number(y), Number(m) - 1, Number(d));
      if (
        date.getFullYear() === Number(y) &&
        date.getMonth() === Number(m) - 1 &&
        date.getDate() === Number(d)
      ) {
        return date;
      }
      return null;
    }
    // ISO fallback
    const iso = new Date(trimmed);
    return isNaN(iso.getTime()) ? null : iso;
  }
  return null;
}

export function today(): Date {
  const d = new Date();
  d.setHours(0, 0, 0, 0);
  return d;
}

export function addYears(date: Date, years: number): Date {
  const d = new Date(date);
  d.setFullYear(d.getFullYear() + years);
  return d;
}

export function dateOnly(d: Date): Date {
  const c = new Date(d);
  c.setHours(0, 0, 0, 0);
  return c;
}

File: backend-java/src/main/java/cl/sernapesca/laboratorio/validators/DateUtils.java
package cl.sernapesca.laboratorio.validators;

import java.time.LocalDate;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.Date;
import java.util.Map;

public class DateUtils {

    private static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("d/M/yyyy");

    public static LocalDate parseDate(Object value) {
        if (value == null) return null;

        if (value instanceof Date) {
            return ((Date) value).toInstant().atZone(ZoneId.systemDefault()).toLocalDate();
        }
        if (value instanceof LocalDate) {
            return (LocalDate) value;
        }
        if (value instanceof Number) {
            // Manejo de fechas seriales de Excel (días desde el 01/01/1900)
            long excelSerial = ((Number) value).longValue();
            long utcDays = excelSerial - 25569;
            return LocalDate.of(1970, 1, 1).plusDays(utcDays);
        }
        if (value instanceof String) {
            String trimmed = ((String) value).trim();
            if (trimmed.isEmpty()) return null;

            // Intentar formato dd/mm/aaaa
            try {
                return LocalDate.parse(trimmed, FORMATTER);
            } catch (DateTimeParseException e) {
                // Fallback a parseo ISO estándar (aaaa-mm-dd)
                try {
                    return LocalDate.parse(trimmed);
                } catch (DateTimeParseException ex) {
                    return null;
                }
            }
        }
        return null;
    }

    public static LocalDate today() {
        return LocalDate.now();
    }

    public static LocalDate addYears(LocalDate date, int years) {
        return date.plusYears(years);
    }
    public static ValidationTypes.RowDisplay buildDisplay(Map<String, Object> rowData) {
        String codigoMuestra = str(rowData.get("codigo_muestra"));
        String nInforme = str(rowData.get("n_informe"));
        Object fMuestreo = rowData.get("fecha_muestreo") != null ? rowData.get("fecha_muestreo") : rowData.get("fecha_elaboracion");
        String fechaMuestreo = fmtDate(fMuestreo);
        
        Object analisisObj = rowData.get("analisis_solicitado");
        if (analisisObj == null || String.valueOf(analisisObj).trim().isEmpty()) {
            analisisObj = rowData.get("tipo_analisis");
        }
        String analisis = str(analisisObj);
        String valorObtenido = str(rowData.get("valor_obtenido"));

        return new ValidationTypes.RowDisplay(codigoMuestra, nInforme, analisis, valorObtenido, fechaMuestreo);
    }

    private static String str(Object v) {
        if (v == null) return "—";
        String s = String.valueOf(v).trim();
        return s.isEmpty() ? "—" : s;
    }

    private static String fmtDate(Object v) {
        if (v == null) return "—";
        if (v instanceof Date d) {
            return new java.text.SimpleDateFormat("dd/MM/yyyy").format(d);
        }
        if (v instanceof LocalDate ld) {
            return ld.format(DateTimeFormatter.ofPattern("dd/MM/yyyy"));
        }
        String s = String.valueOf(v).trim();
        return s.isEmpty() ? "—" : s;
    }
}